python/lsst/daf/butler/registry/interfaces/_datasets.py

```diff
@@ -0,0 +1,409 @@
+        it was initialized or last refreshed.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def find(self, name: str) -> Optional[DatasetRecordStorage]:
+        """Return an object that provides access to the records associated with
+        the given `DatasetType`, if one exists.
+
+        Parameters
+        ----------
+        name : `str`
+            Name of the dataset type.
+
+        Returns
+        -------
+        records : `DatasetRecordStorage` or `None`
+            The object representing the records for the given dataset type, or
+            `None` if there are no records for that dataset type.
+
+        Note
+        ----
+        Dataset types registered by another client of the same repository since
+        the last call to `initialize` or `refresh` may not be found.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def register(self, datasetType: DatasetType) -> Tuple[DatasetRecordStorage, bool]:
+        """Ensure that this `Registry` can hold records for the given
+        `DatasetType`, creating new tables as necessary.
+
+        Parameters
+        ----------
+        datasetType : `DatasetType`
+            Dataset type for which a table should created (as necessary) and
+            an associated `DatasetRecordStorage` returned.
+
+        Returns
+        -------
+        records : `DatasetRecordStorage`
+            The object representing the records for the given dataset type.
+        inserted : `bool`
+            `True` if the dataset type did not exist in the registry before.
+
+        Notes
+        -----
+        This operation may not be invoked within a `Database.transaction`
+        context.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def __iter__(self) -> Iterator[DatasetType]:
+        """Return an iterator over the the dataset types present in this layer.
+
+        Note
+        ----
+        Dataset types registered by another client of the same layer since
+        the last call to `initialize` or `refresh` may not be included.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def getDatasetRef(self, id: int) -> Optional[DatasetRef]:
+        """Return a `DatasetRef` for the given dataset primary key
+        value.
+
+        Parameters
+        ----------
+        id : `int`
+            Autoincrement primary key value for the dataset.
+
+        Returns
+        -------
+        ref : `DatasetRef` or `None`
+            Object representing the dataset, or `None` if no dataset with the
+            given primary key values exists in this layer.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def attachComponents(self, composites: Iterable[Tuple[DatasetRef, Dict[str, DatasetRef]]]
+                         ) -> Iterator[DatasetRef]:
+        """Attach components to one or more datasets.
```

Comment:
It's not 100% clear from documentation but I assume that this is a database operation, not just in-memory update. Could it raise any exceptions?

Reply:
Low-level database exceptions are certainly possible, including unique constraint violations, but I don't think its feasible to document any but the ones that are likely to be common failure modes, and I can't think of any that are likely here.

There are some checks done by the Registry method(s) that call this one, but I prefer to have those in Registry both to not burden all subclasses with performing them and so they can be skipped in internal calls where the checks are unnecessary.